### PR TITLE
outline open statements

### PR DESCRIPTION
**Pull request by vasily-kirichenko**

### CODE REVIEW

File: src/FSharpVSPowerTools.Core/UntypedAstUtils.fs [hunk 1]
```diff
@@ -808,6 +807,32 @@ module Outlining =
...
+                let first = List.rev ranges |> List.head
+                Some (Range.mkRange "" first.Start last.End)
+            | _ -> None
+
+        decls
+        |> List.choose (function SynModuleDecl.Open (_, r) -> Some r | _ -> None)
+        |> List.fold (fun (res, currentRanges) r ->
+            match currentRanges with
+            | [] -> res, [r]
+            | lastRange :: _ when r.StartLine = lastRange.EndLine + 1 -> res, r :: currentRanges
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Core/UntypedAstUtils.fs]: I think the assumption `r.StartLine = lastRange.EndLine + 1` isn't always correct.

I think it maybe simpler to write a custom recursive function that:

- collect all consecutive occurrences of `open` into sublists.
- merge ranges of each sublists.

**Comment by vasily-kirichenko** [comment]
@dungpa done + handle hash (#) directives as well.

File: src/FSharpVSPowerTools.Core/UntypedAstUtils.fs [hunk 2]
```diff
@@ -808,6 +807,34 @@ module Outlining =
...
+                | r :: rest, _ -> loop rest (currentBulk :: res) [r]
+            loop input [] []
+
+
+        decls 
+        |> List.choose predicate
+        |> groupConsecutiveDecls
+        |> List.choose (fun ranges ->
+            match ranges with
+            | [r] when r.StartLine = r.EndLine -> None
```
Comment (dungpa) [inline src/FSharpVSPowerTools.Core/UntypedAstUtils.fs]: Sorry, just to be safe: `[] -> None` case should be added.

Comment (vasily-kirichenko) [inline src/FSharpVSPowerTools.Core/UntypedAstUtils.fs]: Done :)

**Comment by vasily-kirichenko** [comment]
Done.

**Comment by dungpa** [comment]
Great work :+1:

**Pull request merged**